### PR TITLE
v0.4 with rewritten parsers + python bindings

### PR DESCRIPTION
**Pull request by Keats**
Closes DEV-3520, closes DEV-3519, closes DEV-3521.

### CODE REVIEW

**Comment by Keats** [comment]
cc @bovee in case you haven't seen that

**Comment by bovee** [comment]
This is really cool! My only nit is I think your "random TSV" example is actually from a SAM file header, but SAM files are technically TSVs so that's probably okay. :)

The new streaming iterator syntax is neat; I've been playing with building toy parsers with similar style so it's really nice to see needletail has that. Super excited for new stable maturin too; didn't know they were finally nailing that down.

**Comment by luizirber** [comment]
I just tried out this branch (see [this commit](https://github.com/luizirber/decoct/commit/6a04b1a784bfb9edcd91159db1fe5091b252d524)), and was a bit sad to not see a `parse_fastx_reader` function, so I tried out a few things:

- First, using `io::Cursor` and `Box` to avoid the `Seek` bound. Check [this commit](https://github.com/onecodex/needletail/commit/ccb6593af2e590b724d78c736ae11157d2464144). This is a trick I learned with @natir, and we use it in `niffler`. And then...
- [This commit](https://github.com/onecodex/needletail/commit/9a84ed90081fb07ffde7c42c8cb982ec1b2fb163) uses `niffler` to replace the compression detection logic, but implies adding a dependency. I checked the dep tree, and I think the only extra one is `thiserror` and `enum_primitive`. I also added `parse_fastx_reader` as the main method, and `parse_fastx_file` then becomes just a small shim that opens a `File` and passes to `parse_fastx_reader` (still need to fix one `unwarp` and understand how to pass the `compression` feature from `needletail` further down to the right `niffler` feature)

The second commit is pretty aggressive (sorry!) with all the code replacing, but I think the first one is less invasive. I can open a PR against this one if you want.

**Comment by luizirber** [comment]
oh, and I really liked the new iterator style! It makes working with errors a breeze, I used to have a lot of `.expect` in my `parse_sequence_file` function because it didn't allow returning errors...

**Comment by Keats** [comment]
I like the parse_fastx_reader a lot, I forgot to implement it again but it was on my TODO list in the back of my brain.
As for niffler I'm not sure, It's nice to be able to control the decoder. For example you have https://github.com/luizirber/niffler/blob/master/src/lib.rs#L163 which means multi gz file would fail to parse and if I want to change that I would to either fix it in niffle (where it might not be appropriate?) or remove it. We do need the `MultiGzDecoder` here since we do have files in that format (few but some).

File: README.md [hunk 1]
```diff
@@ -8,61 +8,53 @@ Needletail is a MIT-licensed, minimal-copying FASTA/FASTQ parser and _k_-mer pro
 The goal is to write a fast *and* well-tested set of functions that more specialized bioinformatics programs can use.
 Needletail's goal is to be as fast as the [readfq](https://github.com/lh3/readfq) C library at parsing FASTX files and much (i.e. 25 times) faster than equivalent Python implementations at _k_-mer counting.
 
-# Example
+## Example
```
Comment (boydgreenfield) [inline README.md:11]: Is this intentional? Looks a bit like the whole block got commented out in an editor. (Referring not just to the header but to the `rust` code block below too)

File: README.md [hunk 2]
```diff
@@ -71,10 +63,25 @@ git clone https://github.com/onecodex/needletail
...
 Questions are best directed as GitHub issues. We plan to add more documentation soon, but in the meantime "doc" comments are included in the source.
 
-# Contributing
+## Contributing
 
 Please do! We're happy to discuss possible additions and/or accept pull requests.
+
+## Acknowledgements
+Starting from 0.4, the parsers algorithms is taken from [seq_io](https://github.com/markschl/seq_io). While it has been slightly modified, it is mainly
+coming that library. Links to the original files are available in `src/parser/fast{a,q}.rs`.
```
Comment (boydgreenfield) [inline README.md:87]: Typo: "coming from that library".

File: README.md (hunk 1 above)
Comment (Keats) [reply README.md:11]: The # part is intentional, the commented out code not so much...

**Comment by luizirber** [comment]
> As for niffler I'm not sure, It's nice to be able to control the decoder. For example you have https://github.com/luizirber/niffler/blob/master/src/lib.rs#L163 which means multi gz file would fail to parse and if I want to change that I would to either fix it in niffle (where it might not be appropriate?) or remove it. We do need the `MultiGzDecoder` here since we do have files in that format (few but some).

Fair point about controlling your deps. But I also tested changing that `GzDecoder` to a `MultiGzDecoder` and `niffler` worked fine (but need to test with more inputs), so that would be an option too.

**Comment by Keats** [comment]
I'm ok with niffler if it can be changed to MultiGzDecoder, I didn't know if that was ok to change or not as it is pretty niche.

**Comment by luizirber** [comment]
> I'm ok with niffler if it can be changed to MultiGzDecoder, I didn't know if that was ok to change or not as it is pretty niche.

It's niche, but it works fine for regular `gzip` files too (I checked the `flate2` code).
PR opened in niffler: https://github.com/luizirber/niffler/pull/28

**Comment by luizirber** [comment]
niffler 2.1.1 released, with `MultiGzDecoder` as the default decoder for gz files.

**Comment by Keats** [comment]
Cheers, can you make a PR with your changes?

**Comment by Keats** [comment]
@bovee do you have a new email I can put in the Cargo.toml?

**Comment by bovee** [comment]
First initial, last name at gmail. Maybe your email should be in the Cargo.toml instead though since you've fixed up most of the code? ;)

**Comment by Keats** [comment]
I've done both!

**Comment by Keats** [comment]
@luizirber looks like pyo3 is requiring `Send` in its latest version for `pyclass` and the Chain type breaks it. I'm looking into it.

**Pull request merged**